walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -34,23 +35,70 @@
     }
 
     // Seperate relevant TimeRanges from events, put into ArrayList and sort by ascending meeting start time
+    // Optional attendees are handled by running two sets of queries in parallel: a query if we are including optional attendees,
+    // and a query if we are only considering required attendees. If the optional attendee query is empty, then that means that 
+    // there are no valid times if we include optional attendees, and optional attendees are ignored.
     List<TimeRange> attendedMeetings = new ArrayList<>();
+    List<TimeRange> allAttendedMeetings = new ArrayList<>();
     for (Event event : events) {
       // First check if the event in question contains people from request, add those meetings to attendedMeetings
-      Set<String> attendees = new HashSet<>(request.getAttendees());
-      attendees.retainAll(event.getAttendees());
+      // Attendees is only required attendees, allAttendees includes optional attendees
+      Set<String> requiredAttendees = new HashSet<>(request.getAttendees());
+      Set<String> optionalAttendees = new HashSet<>(request.getOptionalAttendees());
+      Set<String> allAttendees = new HashSet<>(Sets.union(requiredAttendees, optionalAttendees));
+      
+      // I couldn't think of a way to factor this, I'm sorry Zu :'(
+      requiredAttendees.retainAll(event.getAttendees());
+      allAttendees.retainAll(event.getAttendees());
 
-      if (!attendees.isEmpty()){
+      if (!requiredAttendees.isEmpty()){
         attendedMeetings.add(event.getWhen());
       }
+
+      if (!allAttendees.isEmpty()) {
+        allAttendedMeetings.add(event.getWhen());
+      }
+
     }
 
     // Sort attendedMeetings so that we can filter out all nested meetings in next step
     Collections.sort(attendedMeetings, TimeRange.ORDER_BY_START);
+    Collections.sort(allAttendedMeetings, TimeRange.ORDER_BY_START);
+
+    // All openings represents the availabilities WITH optional attendees, requiredOpenings represents only required attendees.
+    // if allOpenings is empty, that means that there is a conflict with optional attendees and optional attendees should be ignored.
+    List<TimeRange> requiredOpenings = new ArrayList<>(findOpenings(attendedMeetings, duration));
+    List<TimeRange> allOpenings = new ArrayList<>(findOpenings(allAttendedMeetings, duration));
+
+    return allOpenings.size() == 0 ? requiredOpenings : allOpenings;
```

Comment:
Please add a good comment in the method header on what it returns  ("if there is an opening for eveyone, return it, otherwise return the opening for the required").
In general, this is not the most recommended interface design, because the caller wouldn't (explicitly) know whether the returned openings are for everyone or for the required only. It is recommended to make the interface explicit about what it returns to minimize any potential error/bug. At minimum, it should be explicit in the comment.

Reply:
Made explicit in a comment. I definitely understand your concerns, and will revisit further if you'd like